extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergKeyValuePartitionedLayout.java

```diff
@@ -89,11 +90,18 @@ IcebergTableLocationKey keyFromDataFile(
         final PartitionData partitionData = (PartitionData) dataFile.partition();
         for (final ColumnData colData : outputPartitioningColumns) {
             final String colName = colData.name;
-            final Object colValue = partitionData.get(colData.index);
-            if (colValue != null && !colData.type.isAssignableFrom(colValue.getClass())) {
-                throw new TableDataException("Partitioning column " + colName
-                        + " has type " + colValue.getClass().getName()
-                        + " but expected " + colData.type.getName());
+            final Object colValue;
+            final Object valueFromPartitionData = partitionData.get(colData.index);
+            if (valueFromPartitionData != null) {
+                colValue = IdentityPartitionConverters.convertConstant(
+                        partitionData.getType(colData.index), valueFromPartitionData);
```

Comment:
Is this okay to universally apply? Should we only do it when the partition spec specifies `identity`? https://iceberg.apache.org/spec/#partition-transforms

Reply:
Added a check to ensure identity transform. We would need to test properly for non-identity transform.